src/main/scala/viper/gobra/frontend/Desugar.scala

```diff
@@ -1404,163 +1390,341 @@ object Desugar {
+              addedInvariants = Vector(
+                in.ExprAssertion(in.AtMostCmp(in.Length(visited.op)(src), in.Length(exp)(src))(src))(src),
+                in.ExprAssertion(in.Subset(visited.op, domain)(src))(src))
+
+              dBody = blockD(ctx, info)(body)
+
+              // handle break and continue labels
+              continueLabelName = nm.continueLabel(n, info)
+              continueLoopLabelProxy = in.LabelProxy(continueLabelName)(src)
+              continueLoopLabel = in.Label(continueLoopLabelProxy)(src)
+
+              breakLabelName = nm.breakLabel(n, info)
+              breakLoopLabelProxy = in.LabelProxy(breakLabelName)(src)
+              _ <- declare(breakLoopLabelProxy)
+              breakLoopLabel = in.Label(breakLoopLabelProxy)(src)
+
+              visitedEqDomain = in.Assert(in.ExprAssertion(in.EqCmp(
+                in.SetMinus(visited.op, domain)(src),
+                in.SetLit(keyType, Vector.empty)(src)
+              )(src))(src))(src)
+
+              k = leftOfAssignmentNoInit(shorts(0), info)(keyType)
+              v = if (hasValue) leftOfAssignmentNoInit(shorts(1), info)(valType) else k
+
+              updateKeyVal = in.Seqn(Vector(
+                in.Inhale(in.ExprAssertion(in.And(
+                  in.Contains(k, domain)(src),
+                  in.Negation(in.Contains(k, visited.op)(src))(src))(src))(src))(src)
+              ) ++ (if (hasValue) Vector(in.Inhale(in.ExprAssertion(in.EqCmp(v, in.IndexedExp(exp, k, underlyingType(exp.typ))(src))(src))(src))(src)) else Vector()))(src)
+
+              updateVisited = singleAss(visited, in.Union(visited.op, in.SetLit(keyType, Vector(k))(src))(src))(src)
+
+              ifstmt = in.If(
+                in.EqCmp(in.Length(exp)(src), in.IntLit(0)(src))(src),
+                in.Seqn(
+                  // assert <invariant ...>
+                  (spec.invariants zip dInv).map[in.Stmt]((x: (PExpression, in.Assertion)) => in.Assert(x._2)(meta(x._1, info).createAnnotatedInfo(Source.LoopInvariantNotEstablishedAnnotation))))(src),
+                in.Seqn(
+                  dInvPre ++ dTerPre ++ Vector(
+                    initPerm,
+                    in.While(
+                      in.LessCmp(in.Length(visited.op)(src), in.Length(exp)(src))(src),
+                      dInv ++ addedInvariants, dTer, in.Block(Vector(continueLoopLabelProxy, k) ++ (if (hasValue) Vector(v) else Vector()),
+                    Vector(exhalePerm, updateKeyVal, dBody, continueLoopLabel, inhalePerm, updateVisited) ++ dInvPre ++ dTerPre
+                  )(src))(src), visitedEqDomain, breakLoopLabel
+                ))(src)
+              )(src)
+
+            } yield ifstmt))
+
+          case n@PAssForRange(range, ass, spec, body) if underlyingType(info.typ(range.exp)).isInstanceOf[SliceT] || underlyingType(info.typ(range.exp)).isInstanceOf[MapT] =>
```

Comment:
we are computing `underlyingType` multiple times just to be able to deal with different types in multiple cases. Why not have a single case for `PAssForRange` and then, do, as the first thing, a case analysis on underlyingType of range.exp? What I would even do is to introduce separate methods to handle these cases and just call them from the case analysis.